# Projekt 2 - ELECTRE III i PROMETHEE

#### Filip Marciniak 148148, Szymon Pasternak 148146

# Opis zbioru danych

1. Jaka jest domena problemu? 

W problemie rozważamy koszykarzy z ligi NBA grających w sezonie 2023/2024.

2. Jakie jest źródło danych?

Dane pochodzą z oficjalnej strony NBA - https://www.nba.com/stats/players/traditional.

3. Jaki jest punkt widzenia decydenta?

Jako decydent zależy nam na znalezieniu rankingu wszystkich rozważanych koszykarzy od najlepszego do najgorszego.

4. Jaka jest liczba rozważanych wariantów? Czy w oryginalnym zbiorze danych było ich więcej?

W przygotowanym zbiorze danych jest 20 wariantów. Wybrani zostali koszykarze, który uzyskują najwięcej punktów na mecz oraz zagrali przynajmniej 30 meczów. W oryginalnym zbiorze dostępnych jest 549 koszykarzy.

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.float_format', lambda x: '%.5f' % x)

Poniżej przedstawiony został pełen zbiór danych.

In [2]:
df = pd.read_csv('./Basketball_players.csv', sep=';', nrows=20)
df

PLAYER      PTS      FG%      3P%      FT%      REB  \
0               Joel Embiid 35.30000 53.30000 36.60000 88.30000 11.30000   
1               Luka Doncic 33.90000 48.70000 37.40000 78.70000  9.10000   
2     Giannis Antetokounmpo 30.70000 61.60000 28.80000 65.70000 11.50000   
3   Shai Gilgeous-Alexander 30.40000 54.00000 36.70000 87.10000  5.60000   
4              Kevin Durant 27.60000 52.90000 42.20000 85.50000  6.80000   
5          Donovan Mitchell 27.40000 46.80000 37.30000 86.20000  5.30000   
6             Jalen Brunson 27.40000 47.70000 40.10000 84.50000  3.60000   
7              Jayson Tatum 27.30000 47.20000 37.80000 83.20000  8.20000   
8              Devin Booker 27.20000 49.20000 36.40000 88.10000  4.80000   
9             Stephen Curry 26.50000 44.70000 40.30000 92.30000  4.40000   
10             De'Aaron Fox 26.50000 46.60000 36.30000 73.00000  4.50000   
11               Trae Young 26.40000 42.60000 37.10000 85.60000  2.70000   
12          Anthony Edwards 26.20000 46.40000 37.30000 82.70000  5.50000   
13             Nikola Jokic 26.10000 58.10000 35.00000 82.80000 12.20000   
14             Tyrese Maxey 25.80000 44.90000 37.90000 86.70000  3.70000   
15             LeBron James 25.30000 53.00000 40.50000 74.80000  7.30000   
16             Kyrie Irving 25.30000 49.10000 40.70000 90.50000  5.00000   
17            Anthony Davis 24.70000 55.30000 28.90000 81.50000 12.60000   
18           Damian Lillard 24.40000 42.60000 35.80000 92.30000  4.40000   
19            Julius Randle 24.00000 47.20000 31.10000 78.10000  9.20000   

        AST     TOV      PF      +/-  
0   5.70000 3.70000 2.90000  6.90000  
1   9.80000 4.00000 2.10000  3.50000  
2   6.40000 3.40000 2.90000  4.60000  
3   6.30000 2.20000 2.50000  7.90000  
4   5.20000 3.30000 1.70000  3.80000  
5   6.10000 2.90000 2.20000  7.00000  
6   6.50000 2.40000 1.90000  6.10000  
7   4.90000 2.50000 2.00000  8.80000  
8   6.90000 2.50000 3.00000  4.60000  
9   5.00000 2.80000 1.60000  2.00000  
10  5.60000 2.70000 2.60000  2.60000  
11 10.80000 4.30000 2.10000 -2.20000  
12  5.20000 3.10000 1.80000  6.50000  
13  9.00000 2.90000 2.50000  8.20000  
14  6.10000 1.70000 2.20000  3.50000  
15  8.10000 3.30000 1.20000  2.70000  
16  5.30000 1.70000 1.90000  3.70000  
17  3.60000 2.00000 2.40000  1.10000  
18  7.10000 2.50000 1.80000  5.80000  
19  5.00000 3.50000 2.70000  4.70000

5. Opisz jeden z rozważanych wariantów - jak jest oceniony na poszczególnych kryteriach, jak dobry jest na tle innych.

**Nikola Jokic**:
- PTS - 26.1 - znajduje się w drugim kwartale, nie wypada najlepiej, ale nie jest też najgorzej,
- FG% - 58.1 - na tym kryterium jest jednym z najlepszych wariantów,
- 3P% - 35.0 - na tym kryterium znajduje się w pierwszym kwartale,
- FT% - 82.2 - znajduje się w drugim kwartale,
- REB - 12.2 - wyróżnia się spośród innych wariantów dwucyfrowym wynikiem na tym kryterium, są jedynie 4 oceny dwucyfrowe,
- AST - 9.0 - na tym kryterium również jest jednym z lepszych wariantów,
- TOV - 2.9 - na tym kryterium, radzi sobie gorzej niż inne warianty, ale o niewielką wartość
- PF - 2.5 - na tym kryterium radzi sobie średnio,
- +/- - 8.2 - ostatnie kryterium zdecydowanie wyróżnia go spośród innych wariantów, uzyskuje na nim jedną z najlepszych ocen.

6. Ile rozważasz kryteriów? Czy było ich więcej w oryginalnym zbiorze danych?
   
W problemie rozważanych jest 9 kryteriów. W oryginalnym zbiorze danych było ich więcej natomiast wybrane zostały takie, które uśredniają niektóre z dostępnych oryginalnie, np. 3P% lub są najbardziej istotne, takie jak np. liczba punktów na mecz.

8. Jakie są typy poszczególnych kryteriów? Scharakteryzuj je:
Na wszystkich z kryteriów uzyskane wartości są ciągłe, ponieważ są to statystyki średnie na mecz.

- PTS - ZYSK - średnia liczba punktów na mecz
- FG% - ZYSK - procent trafionych rzutów
- 3P% - ZYSK - procent trafionych rzutów 3 punktowych
- FT% - ZYSK - procent trafionych rzutów osobistych
- REB - ZYSK - średnia liczba zbiórek na mecz
- AST - ZYSK - średnia liczba asyst na mecz
- TOV - KOSZT - średnia liczba strat na mecz
- PF - KOSZT - średnia liczba fauli na mecz
- +/- - ZYSK - średnia różnica pomiędzy punktami zdobytymi i straconymi przez drużynę w czasie kiedy gracz był obecny na boisku na mecz

Osiągane wartości:

In [3]:
df.describe()

PTS      FG%      3P%      FT%      REB      AST      TOV       PF  \
count 20.00000 20.00000 20.00000 20.00000 20.00000 20.00000 20.00000 20.00000   
mean  27.42000 49.59500 36.71000 83.38000  6.88500  6.43000  2.87000  2.20000   
std    2.98780  5.08749  3.61137  6.68026  3.11487  1.78771  0.71532  0.48232   
min   24.00000 42.60000 28.80000 65.70000  2.70000  3.60000  1.70000  1.20000   
25%   25.67500 46.55000 36.17500 80.80000  4.47500  5.20000  2.47500  1.87500   
50%   26.50000 48.20000 37.20000 85.00000  5.55000  6.10000  2.85000  2.15000   
75%   27.45000 53.07500 38.45000 87.35000  9.12500  6.95000  3.32500  2.52500   
max   35.30000 61.60000 42.20000 92.30000 12.60000 10.80000  4.30000  3.00000   

           +/-  
count 20.00000  
mean   4.59000  
std    2.68110  
min   -2.20000  
25%    3.30000  
50%    4.60000  
75%    6.60000  
max    8.80000

8. Jakie są wagi poszczególnych kryteriów?

Aby ustalić wagi danych kryteriów zastosowana została procedura SRF. Poniżej zaprezentowane zostało ustawienie kart:

<div>
    <img src="./criteria_cards.png" alt="cards" width="700"/>
</div>

`Ratio Z` ustawione zostało na 6. Po wykonaniu obliczeń wagi przedstawiają się nastepująco:

| CRITERION | WEIGHT |
| --------- | ------ |
| PTS | 6 |
| AST, REB | 3.86 |
| FG%, 3P% | 3.14 |
| TOV | 2.43 |
| FT%, +/- | 1.71 |
| PF | 1 |

In [45]:
CRITERIA = {
    'PTS': {'type': 'gain', 'k': 6, 'q': 4, 'p': 8, 'v': 12},
    'FG%': {'type': 'gain', 'k': 3.14, 'q': 3, 'p': 4, 'v': 8},
    '3P%': {'type': 'gain', 'k': 3.14, 'q': 4, 'p': 5, 'v': 11},
    'FT%': {'type': 'gain', 'k': 1.71, 'q': 3, 'p': 7, 'v': 15},
    'REB': {'type': 'gain', 'k': 3.86, 'q': 2, 'p': 5, 'v': 9},
    'AST': {'type': 'gain', 'k': 3.86, 'q': 1.5, 'p': 3, 'v': 8},
    'TOV': {'type': 'cost', 'k': 2.43, 'q': 1, 'p': 1.5, 'v': 2.5},
    'PF':  {'type': 'cost', 'k': 1, 'q': 1, 'p': 1.5, 'v': 2},
    '+/-': {'type': 'gain', 'k': 1.71, 'q': 3, 'p': 4, 'v': 8},
}

9. Czy istnieją jakieś dominacje w zbiorze wariantów? Jeśli tak, podaj je.

W zbiorze istnieją warianty zdominowane przez inne.

In [5]:
for alt_1_id in df.index:
    for alt_2_id in df.index:
        if alt_1_id == alt_2_id:
            continue

        is_at_least_as_good_on_all = all([df.loc[alt_1_id, cr] <= df.loc[alt_2_id, cr] for cr, config in CRITERIA.items() if config['type'] == 'gain']) \
               and all([df.loc[alt_1_id, cr] >= df.loc[alt_2_id, cr] for cr, config in CRITERIA.items() if config['type'] == 'cost'])
        is_better_on_one = any([df.loc[alt_1_id, cr] < df.loc[alt_2_id, cr] for cr, config in CRITERIA.items() if config['type'] == 'gain']) \
                or any([df.loc[alt_1_id, cr] > df.loc[alt_2_id, cr] for cr, config in CRITERIA.items() if config['type'] == 'cost'])
        
        if is_at_least_as_good_on_all and is_better_on_one:
            print(f"{df.loc[alt_2_id, 'PLAYER'].ljust(23)} dominates {df.loc[alt_1_id, 'PLAYER']}")

Shai Gilgeous-Alexander dominates De'Aaron Fox
Nikola Jokic            dominates Julius Randle


10. Jak teoretycznie powinien prezentować się najlepszy wariant w Twoim zbiorze danych?

Najlepszy wariant powinien osiągać bardzo dobre wartości, przede wszystkim, na kryteriach, które mają największą wagę. Wydaje się, że najbardziej wszechstronnnym graczem na kryteriach PTS, AST, REB, FG%, 3P% jest Luka Doncic lub Nikola Jokic. Każdy z nich jest jednak poniżej liderów na jednym z ważnych kryteriów. Dla Luki Doncica jest to kryterium FG%, a dla Nikoli Jokica PTS. Najlepszy gracz powinien mieć wyrównane, dobre statystyki na większości kryteriów, ponieważ wagi nie różnią się znacząco. Gdyby jednak jego statystyki na jednym z kryteriów odstawały na jego niekorzyść od reszty, to lepiej żeby to było kryterium o mniejszej wadze. W przykładzie powyżej, lepiej prezentuje się Nikola Jokic, bo pomimo niskiej wartości na kryterium PTS, które posiada najwyższą wagę, jego pozostałe statystyki (w szczególności AST i REB) prezentują się bardzo dobrze.

11. Spośród rozważanych wariantów, który wydaje się być najlepszy/lepszy od innych i dlaczego?

Najlepsi będą koszykarze, którzy mają najwięcej zdobytych punktów na mecz, ponieważ to kryterium posiada największą wagę. Najlepsi więc wydają się być Joel Embiid i Luka Doncic.

12. Spośród rozważanych wariantów, który wydaje się być najsłabszy/ znacząco słabszy od innych i dlaczego?

Najgorsi będą koszykarze, którzy mają najmniej zdobytych punktów na mecz, ponieważ to kryterium posiada największą wagę. Najgorszy więc wydaje się być Julius Randle.

**Porównania parami do interpretacji wyników**

- Nikola Jokic > Stephen Curry,
- Joel Embiid > Kevin Durant,
- Nikola Jokic ~ Luka Doncic,
- Jason Tatum > Anthony Edwards

# ELECTRE III

1. Informacja preferencyjna podana na wejście metody:

| KRYTERIUM | Próg nierozróżnialności (q) | Próg preferencji (p) | Próg veta (v) |
| --------- | --------------------------- | -------------------- | ------------- |
| PTS | 3 | 5 | 8 |
| FG% | 3 | 4 | 8 |
| 3P% | 4 | 5 | 11 |
| FT% | 4 | 8 | 13 |
| REB | 2 | 5 | 9 |
| AST | 2 | 4 | 7 |
| TOV | 1 | 1.5 | 2 |
| PF  | 1 | 1.5 | 2 |
| +/- | 3 | 4 | 8 |

In [28]:
CRITERIA = {
    'PTS': {'type': 'gain', 'k': 6,    'q': 3,   'p': 5,   'v': 8},
    'FG%': {'type': 'gain', 'k': 3.14, 'q': 3,   'p': 4,   'v': 8},
    '3P%': {'type': 'gain', 'k': 3.14, 'q': 4,   'p': 5,   'v': 11},
    'FT%': {'type': 'gain', 'k': 1.71, 'q': 4,   'p': 8,   'v': 13},
    'REB': {'type': 'gain', 'k': 3.86, 'q': 2,   'p': 5,   'v': 9},
    'AST': {'type': 'gain', 'k': 3.86, 'q': 2,   'p': 4,   'v': 7},
    'TOV': {'type': 'cost', 'k': 2.43, 'q': 1,   'p': 1.5, 'v': 2},
    'PF':  {'type': 'cost', 'k': 1,    'q': 1,   'p': 1.5, 'v': 2},
    '+/-': {'type': 'gain', 'k': 1.71, 'q': 3,   'p': 4,   'v': 8},
}

## Marginal concordance and discordance

In [29]:
def calculate_marginal_concordance(ga, gb, q, p, type_='gain'):

    if type_ == 'gain':
        if ga - gb >= -q:
            return 1
        elif ga - gb < -p:
            return 0
        else:
            return (p-(gb-ga))/(p-q)
    elif type_ == 'cost':
        if ga - gb <= q:
            return 1
        elif ga - gb > p:
            return 0
        else:
            return (p-(ga-gb))/(p-q)
    else:
        raise Exception("wrong criterion type")
    

def calculate_marginal_discordance(ga, gb, p, v, type_='gain'):

    if type_ == 'gain':
        if ga - gb <= -v:
            return 1
        elif ga - gb >= -p:
            return 0
        else:
            return ((gb-ga)-p)/(v-p)
    elif type_ == 'cost':
        if ga - gb >= v:
            return 1
        elif ga - gb <= p:
            return 0
        else:
            return (v-(ga-gb))/(v-p)
    else:
        raise Exception("wrong criterion type")

## Comprehensive Concordance

In [30]:
# Comprehensive Concordance will be stored inside df_C variable which is a NxN dataframe
df_C = pd.DataFrame(1.0, index=df.index, columns=df.index)
weights = sum([config['k'] for config in CRITERIA.values()])

for alt_1_id in df_C.index:
    for alt_2_id in df_C.index:
        if alt_1_id == alt_2_id:
            continue

        comprehensive_concordance_index = 0
        # calculate marginal concordance index for each criterion and add it to the cc index with appropriate weight
        for criterion, config in CRITERIA.items():
            marginal_concordance = calculate_marginal_concordance(
                ga=df.loc[alt_1_id, criterion],
                gb=df.loc[alt_2_id, criterion],
                q=config['q'],
                p=config['p'],
                type_=config['type']
            )
            comprehensive_concordance_index += config['k'] * marginal_concordance
        # normalize the cc index
        comprehensive_concordance_index /= weights
        df_C.loc[alt_1_id, alt_2_id] = comprehensive_concordance_index

## Outranking Credibility

In [31]:
df_sigma = pd.DataFrame(1.0, index=df.index, columns=df.index)

for alt_1_id in df_C.index:
    for alt_2_id in df_C.index:
        if alt_1_id == alt_2_id:
            continue

        C = sigma = df_C.loc[alt_1_id, alt_2_id]
        # calculate marginal discordance for each criterion
        for criterion, config in CRITERIA.items():
            marginal_discordance = calculate_marginal_discordance(
                ga=df.loc[alt_1_id, criterion],
                gb=df.loc[alt_2_id, criterion],
                p=config['p'],
                v=config['v'],
                type_=config['type']
            )
            # if the marginal discordance is greater than the result of concordance test,
            # take it into account while calculating outranking credibility (sigma)
            if marginal_discordance > C:
                sigma = sigma * (1-marginal_discordance)/(1-C)
        df_sigma.loc[alt_1_id, alt_2_id] = sigma

np.fill_diagonal(df_sigma.values, -1)

## Distillation

In [32]:
def distillation(df_m, upper_threshold=None, descending=True):

    # compute upper threshold
    if upper_threshold is None:
        upper_threshold = df_m.values.max()
    
    # compute lower threshold
    s = -0.15 * upper_threshold + 0.3
    lower_threshold = upper_threshold - s
    lower_threshold = df_m[(df_m < lower_threshold) & (df_m > -1)].max().max()
    if np.isnan(lower_threshold):
        lower_threshold = 0
    
    # print(df_m)
    # print(f"{s=}, {lower_threshold=}, {type(lower_threshold)} {upper_threshold=}")

    # if upper threshold is equal to 0, then stop the algorithm
    if upper_threshold == 0:
        return df_m.index

    # save only relations that satisfy given conditions to the df_f (df filtered)
    df_f = pd.DataFrame(-1.0, index=df_m.index, columns=df_m.index)
    for alt_1_id in df_m.index:
        for alt_2_id in df_m.columns:
            if alt_1_id == alt_2_id:
                continue
            sigma = df_m.loc[alt_1_id, alt_2_id]
            if sigma > lower_threshold and sigma > df_m.loc[alt_2_id, alt_1_id] + s:
                df_f.loc[alt_1_id, alt_2_id] = sigma

    # compute the strength, weakness and quality
    df_swq = pd.DataFrame(0, index=df_m.index, columns=['strength', 'weakness', 'quality'])
    df_swq['strength'] = df_f.apply(lambda row: (row != -1.0).sum(), axis=1)
    df_swq['weakness'] = df_f.apply(lambda col: (col != -1.0).sum(), axis=0)
    df_swq['quality'] = df_swq['strength'] - df_swq['weakness']

    # take subset with maximal quality
    df_swq['rank'] = df_swq['quality'].rank(ascending=False, method='min')
    if descending:
        df_mq = df_swq[df_swq['rank'] == 1]
    else:
        df_mq = df_swq[df_swq['rank'] == df_swq['rank'].max()]
    if len(df_mq.index) == 1:
        return df_mq.index
    else:
        return distillation(
            df_m.loc[df_mq.index, df_mq.index],
            upper_threshold=lower_threshold,
            descending=descending
        )

## Adding descending preorder and ascending preorder to the main dataframe

In [33]:
df['desc_preorder'] = 0
df['asc_preorder'] = 0

descending_results = []
df_copy = df_sigma.copy()
position = 1
while len(df_copy.index) > 0:
    alternatives = distillation(df_copy, descending=True)
    df_copy = df_copy.loc[~df_copy.index.isin(alternatives), ~df_copy.index.isin(alternatives)]
    df.loc[alternatives, 'desc_preorder'] = position
    descending_results.append([df.loc[alt, "PLAYER"] for alt in alternatives])
    position += 1
print(f"{descending_results=}")

ascending_results = []
df_copy = df_sigma.copy()
position = len(df_sigma)
while len(df_copy.index) > 0:
    alternatives = distillation(df_copy, descending=False)
    df_copy = df_copy.loc[~df_copy.index.isin(alternatives), ~df_copy.index.isin(alternatives)]
    df.loc[alternatives, 'asc_preorder'] = position
    ascending_results.insert(0, [df.loc[alt, "PLAYER"] for alt in alternatives])
    position -= 1
df['asc_preorder'] = df['asc_preorder'] - df['asc_preorder'].min() + 1
print(f"{ascending_results=}")

descending_results=[['Joel Embiid'], ['Nikola Jokic'], ['Shai Gilgeous-Alexander'], ['Kevin Durant'], ['Luka Doncic'], ['Jayson Tatum', 'Kyrie Irving'], ['LeBron James'], ['Anthony Davis'], ['Donovan Mitchell', 'Stephen Curry', 'Tyrese Maxey', 'Damian Lillard'], ['Giannis Antetokounmpo', 'Jalen Brunson', 'Devin Booker', 'Anthony Edwards'], ['Trae Young'], ["De'Aaron Fox", 'Julius Randle']]
ascending_results=[['Joel Embiid', 'Luka Doncic'], ['Shai Gilgeous-Alexander', 'Nikola Jokic'], ['Giannis Antetokounmpo'], ['Kyrie Irving'], ['Kevin Durant', 'Jayson Tatum'], ['Stephen Curry', 'Anthony Davis'], ['Donovan Mitchell', 'Jalen Brunson', 'Devin Booker'], ['LeBron James'], ['Tyrese Maxey'], ['Anthony Edwards', 'Damian Lillard'], ['Julius Randle'], ['Trae Young'], ["De'Aaron Fox"]]


## Creating df_final with final graph

Final preorder is the intersection of descending and ascending preorders

In [39]:
df_final = pd.DataFrame(0, index=df.index, columns=df.index)

for alt_1_id in df_final.index:
    for alt_2_id in df_final.columns:
        if alt_1_id == alt_2_id:
            continue

        alt_1_d = df.loc[alt_1_id, "desc_preorder"]
        alt_1_a = df.loc[alt_1_id, "asc_preorder"]
        alt_2_d = df.loc[alt_2_id, "desc_preorder"]
        alt_2_a = df.loc[alt_2_id, "asc_preorder"]

        if alt_1_d < alt_2_d and alt_1_a < alt_2_a:
            df_final.loc[alt_1_id, alt_2_id] = 1
        elif (alt_1_d < alt_2_d and alt_1_a == alt_2_a) or (alt_1_d == alt_2_d and alt_1_a < alt_2_a):
            df_final.loc[alt_1_id, alt_2_id] = 1
        elif alt_1_d == alt_2_d and alt_1_a == alt_2_a:
            df_final.loc[alt_1_id, alt_2_id] = 2

# ---------------------------------------------------------------------------------------------------- #
# Code used to produce output for graph visualization
            
# relations = []
# for alt_1_id in df_final.index:
#     for alt_2_id in df_final.columns:
#         if alt_1_id == alt_2_id:
#             continue
#         if df_final.loc[alt_1_id, alt_2_id]:
#             relations.append({"alternative_1": alt_1_id, "alternative_2": alt_2_id})

# alternatives = []
# for alt_1_id in df.index:
#     alternatives.append({"id": alt_1_id, "name": df.loc[alt_1_id, "PLAYER"]})

# print(relations)
# print(alternatives)

## Finding rank for each alternative

In [40]:
def find_longest_paths(df):
    def dfs(node, path):
        if node not in parents:
            return
        for parent in parents[node]:
            dfs(parent, path + [parent])
        longest_paths[node] = max(longest_paths[node], path, key=len)

    # find parents for each node
    parents = {}
    for row in df.index:
        parents[row] = [col for col in df.columns if df.loc[row, col] == 1]

    # initialize path for each node with empty list
    longest_paths = {node: [] for node in df.columns}

    # perform DFS from each node
    for node in df.columns:
        dfs(node, [])

    return longest_paths

longest_paths = find_longest_paths(df_final)
df['ranks'] = [len(path)+1 for path in longest_paths.values()]

## Finding median rank for each alternative

In [42]:
df['median'] = 0
position = 1
for rank_value in df['ranks'].sort_values().unique():

    rank_group = df[df['ranks'] == rank_value].copy()
    rank_group['plus'] = rank_group['desc_preorder'] + rank_group['asc_preorder']
    rank_group = rank_group.sort_values('plus')
    for unique_plus in rank_group['plus'].unique():
        rank_group_unique_plus = rank_group[rank_group['plus'] == unique_plus]
        for alt_id in rank_group_unique_plus.index:
            df.loc[alt_id, "median"] = position

        position += 1

# Final ranking and median ranking

In [46]:
df[['PLAYER', 'desc_preorder', 'asc_preorder', 'ranks', 'median']].sort_values(['ranks', 'median'])

The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.


PLAYER  desc_preorder  asc_preorder  ranks  median
0               Joel Embiid              1             1      1       1
13             Nikola Jokic              2             2      2       2
1               Luka Doncic              5             1      2       3
3   Shai Gilgeous-Alexander              3             2      3       4
4              Kevin Durant              4             5      4       5
16             Kyrie Irving              6             4      4       6
2     Giannis Antetokounmpo             10             3      4       7
7              Jayson Tatum              6             5      5       8
17            Anthony Davis              8             6      6       9
15             LeBron James              7             8      6      10
9             Stephen Curry              9             6      7      11
5          Donovan Mitchell              9             7      8      12
6             Jalen Brunson             10             7      9      13
8              Devin Booker             10             7      9      13
14             Tyrese Maxey              9             9      9      14
18           Damian Lillard              9            10     10      15
12          Anthony Edwards             10            10     11      16
11               Trae Young             11            12     12      17
19            Julius Randle             12            11     12      17
10             De'Aaron Fox             12            13     13      18

2. Powyżej zaprezentowana jest tabela przedstawiająca ranking końcowy (kolumna ranks) oraz ranking medianowy (kolumna median).
3. Otrzymywane wyniki zgodne są z oczekiwaniami. Najlepiej wypadł Joel Embiid oraz Nikola Jokic. Najgorzej natomiast De'Aaron Fox oraz Julius Randle. Można zauważyć że największą wagę odegrała liczba zdobytych punktów, natomiast nie zawsze było to najważniejsze czego dobrym przykładem jest Nikola Jokic. Porównując otrzymane wyniki z ustalonymi a priori porównaniami pomiędzy wariantami można zaobserwować, że wszystkie są spełnione, natomiast `Nikola Jokic ~ Luka Doncic` jest spełnione tylko dla rang. W porządku medianowym to Nikola Jokic okazuje się być lepszy ze względu na słabszą pozycję Doncica w porządku zstępującym, więc nierozróżnialność nie jest zachowana. Co ciekawe nawet dla rankingu medianowego zachodzi nierozróżnialność dla pary wariantów Jalen Brunson i Devin Brooker.

# PROMETHEE

Informacja preferencyjna podana na wejście metody

| KRYTERIUM | Próg nierozróżnialności (q) | Próg preferencji (p) |
| --------- | --------------------------- | -------------------- |
| PTS | 3 | 5 |
| FG% | 3 | 4 |
| 3P% | 4 | 5 |
| FT% | 4 | 8 |
| REB | 2 | 5 |
| AST | 2 | 4 |
| TOV | 1 | 1.5 |
| PF  | 1 | 1.5 |
| +/- | 3 | 4 |

In [56]:
PROMETHEE = {
    'PTS': {'type': 'gain', 'k': 6,    'q': 3,   'p': 5},
    'FG%': {'type': 'gain', 'k': 3.14, 'q': 3,   'p': 4},
    '3P%': {'type': 'gain', 'k': 3.14, 'q': 4,   'p': 5},
    'FT%': {'type': 'gain', 'k': 1.71, 'q': 4,   'p': 8},
    'REB': {'type': 'gain', 'k': 3.86, 'q': 2,   'p': 5},
    'AST': {'type': 'gain', 'k': 3.86, 'q': 2,   'p': 4},
    'TOV': {'type': 'cost', 'k': 2.43, 'q': 1,   'p': 1.5},
    'PF':  {'type': 'cost', 'k': 1,    'q': 1,   'p': 1.5},
    '+/-': {'type': 'gain', 'k': 1.71, 'q': 3,   'p': 4},
}

In [57]:
def calculate_marginal_preference(ga, gb, q, p, type_='gain'):
    if type_ == 'gain':
        if ga - gb > p:
            return 1
        elif ga - gb <= q:
            return 0
        else:
            return ((ga-gb)-q)/(p-q)
    elif type_ == 'cost':
        if gb - ga > p:
            return 1
        elif gb - ga <= q:
            return 0
        else:
            return ((gb-ga)-q)/(p-q)
    else:
        raise Exception("wrong criterion type")

## Comprehensive Preference Index

In [58]:
# Comprehensive Preference Index will be stored inside df_C variable which is a NxN dataframe
df_PI = pd.DataFrame(1.0, index=df.index, columns=df.index)
weights = sum([config['k'] for config in PROMETHEE.values()])

for alt_1_id in df_PI.index:
    for alt_2_id in df_PI.index:
        if alt_1_id == alt_2_id:
            continue

        comprehensive_preference_index = 0
        # calculate marginal preference index for each criterion and add it to the cp index with appropriate weight
        for criterion, config in PROMETHEE.items():
            marginal_preference = calculate_marginal_preference(
                ga=df.loc[alt_1_id, criterion],
                gb=df.loc[alt_2_id, criterion],
                q=config['q'],
                p=config['p'],
                type_=config['type']
            )
            comprehensive_preference_index += config['k'] * marginal_preference
        # normalize the cp index
        comprehensive_preference_index /= weights
        df_PI.loc[alt_1_id, alt_2_id] = comprehensive_preference_index

## Calculating Positive and Negative Flows and PROMETHEE rankings

In [59]:
# calculating flows
df['P_negative_flow'] = df_PI.sum()
df['P_positive_flow'] = df_PI.sum(axis=1)
df['P_balance'] = df['P_positive_flow'] - df['P_negative_flow']

# calculating negative and positive ranking
df['P_ranking_negative'] = df['P_negative_flow'].rank(method='min').astype(int)
df['P_ranking_positive'] = df['P_positive_flow'].rank(ascending=False, method='min').astype(int)

df_ranking_promethee_I = pd.DataFrame(0, index=df.index, columns=df.index)
for alt_1_id in df_ranking_promethee_I.index:
    for alt_2_id in df_ranking_promethee_I.columns:
        if alt_1_id == alt_2_id:
            continue

        alt_1_n = df.loc[alt_1_id, 'P_negative_flow']
        alt_1_p = df.loc[alt_1_id, 'P_positive_flow']
        alt_2_n = df.loc[alt_2_id, 'P_negative_flow']
        alt_2_p = df.loc[alt_2_id, 'P_positive_flow']
        # preference
        if alt_1_p >= alt_2_p and alt_1_n <= alt_2_n:
            df_ranking_promethee_I.loc[alt_1_id, alt_2_id] = 1

df_ranking_promethee_II = pd.DataFrame(0, index=df.index, columns=df.index)
for alt_1_id in df_ranking_promethee_II.index:
    for alt_2_id in df_ranking_promethee_II.columns:
        if alt_1_id == alt_2_id:
            continue

        if df.loc[alt_1_id, 'P_balance'] >= df.loc[alt_2_id, 'P_balance']:
            df_ranking_promethee_II.loc[alt_1_id, alt_2_id] = 1

# ---------------------------------------------------------------------------------------------------- #
# Code used to produce output for graph visualization

# PROMETHEE I
# print(alternatives)

# relations = []
# for alt_1_id in df_ranking_promethee_I.index:
#     for alt_2_id in df_ranking_promethee_I.columns:
#         if alt_1_id == alt_2_id:
#             continue
#         if df_ranking_promethee_I.loc[alt_1_id, alt_2_id]:
#             relations.append({"alternative_1": alt_1_id, "alternative_2": alt_2_id})

# alternatives = []
# for alt_1_id in df.index:
#     alternatives.append({"id": alt_1_id, "name": df.loc[alt_1_id, "PLAYER"]})

# print(relations)

# # PROMETHEE II
# relations = []
# for alt_1_id in df_ranking_promethee_II.index:
#     for alt_2_id in df_ranking_promethee_II.columns:
#         if alt_1_id == alt_2_id:
#             continue
#         if df_ranking_promethee_II.loc[alt_1_id, alt_2_id]:
#             relations.append({"alternative_1": alt_1_id, "alternative_2": alt_2_id})

# print(relations)

## Ranking częściowy (Promethee I)

<div>
    <img src="./promethee_I.svg" alt="promethee_I" width="700"/>
</div>

## Ranking całkowity (Promethee II)

<div>
    <img src="./promethee_II.svg" alt="promethee_II" width="700" height="500"/>
</div>

## Porównanie rankingów otrzymanych przy pomocy Promethee I i Promethee II

Ranking otrzymany metodą Promethee I dopuszcza nieporównywalności wariantów co widoczne jest na zamieszczonym wyżej diagramie. Powoduje to, że trudniej jest odczytać z niego faktyczne uporządkowanie wariantów, a możliwe jest jedynie wyodrębnienie poszczególnych grup wariantów na podstawie poziomów na których znajdują się w grafie.
Ranking otrzymany metodą Promethee II jest zupełny dzięki czemu jego interpretacja jest łatwiejsza. Ogólna struktura rankingu jest zachowana w obu metodach: najlepiej wypadają Joel Embiid, Luka Doncic i Nikola Jokic, a najgorzej De'Aaron Fox oraz Julius Randle (mimo tego, że w Promethee I nie znajduje się na dole diagramu, to nie jest lepszy od żadnego innego wariantu).

## Porównanie wyników z oczekiwaniami i preferencjami

Dla Promethee I wszystkie ustalone preferencje są spełnione, a dla Promethee II nie jest spełnione jedynie porównanie `Luka Doncic ~ Nikola Jokic`. W metodzie Promethee II trudno jednak uzyskać nierozróżnialność dwóch wariantów, ponieważ przepływy całkowite musiałyby być sobie równe, co jest trudne do osiągnięcia.

# Porównanie wyników ELECTRE i PROMETHEE

Dostarczony przez metodę ELECTRE III ranking medianowy jest najbardziej zgodny z oczekiwaniami. Wystepuje też w nim jedna nierozróżnialność. Przy pomocy metody PROMETHEE uzyskano dwa wyniki - z metody PROMETHEE I i II. Wynik metody PROMETHEE jest trudno interpretowalny ze względu na dużą liczbę nieporównywalności pomiędzy wariantami. PROMETHEE II dostarcza jednak ranking zupełny, który łatwiej zinterpretować. Wyniki dla wszystkich metod są dość podobne szczególnie jeżeli chodzi o warianty na czołowych pozycjach (Joel Embiid, Nikola Jokic i Luka Doncic). Różnice występują w środku rankingów. Przykładowo Stephen Curry zajmuje 11 miejsce w ELECTRE III, a dopiero 16 w PROMETHEE II. W ogólności wyniki są jednak dość podobne.